src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_secrets.rs

```diff
@@ -37,6 +50,186 @@ pub fn is_base64_url_encoding_char(ch: char) -> bool
+    let actual_checksum_bytes = &key_bytes[key_bytes_length - 6..key_bytes_length - 3];
+
+    let computed_marvin = marvin::compute_hash32(bytes_for_checksum, checksum_seed, 0, bytes_for_checksum.len() as i32);
+    let computed_marvin_bytes = computed_marvin.to_ne_bytes();
+
+    // The HIS v2 standard requires a match for the first 3-bytes (24 bits) of the Marvin checksum.
+    actual_checksum_bytes[0] == computed_marvin_bytes[0]
+    && actual_checksum_bytes[1] == computed_marvin_bytes[1]
+    && actual_checksum_bytes[2] == computed_marvin_bytes[2]
+}
+
+pub fn generate_common_annotated_key(base64_encoded_signature: &str,
+    customer_managed_key: bool,
+    platform_reserved: Option<&[u8]>,
+    provider_reserved: Option<&[u8]>,
+    test_char: Option<char>) -> Result<String, String> {
+generate_common_annotated_test_key(VERSION_TWO_CHECKSUM_SEED.clone(),
+      base64_encoded_signature,
+      customer_managed_key,
+      platform_reserved,
+      provider_reserved,
+      test_char)
+}
+
+pub fn generate_common_annotated_test_key(
+    checksum_seed: u64,
+    base64_encoded_signature: &str,
+    customer_managed_key: bool,
+    platform_reserved: Option<&[u8]>,
+    provider_reserved: Option<&[u8]>,
+    test_char: Option<char>,
+) -> Result<String, String> {
+    const PLATFORM_RESERVED_LENGTH: usize = 9;
+    const PROVIDER_RESERVED_LENGTH: usize = 3;
+
+    let platform_reserved = match platform_reserved {
+        Some(reserved) if reserved.len() != PLATFORM_RESERVED_LENGTH => {
+            return Err(format!(
+                "When provided, there must be {} reserved bytes for platform metadata.",
+                PLATFORM_RESERVED_LENGTH
+            ));
+        }
+        Some(reserved) => reserved,
+        None => &[0; PLATFORM_RESERVED_LENGTH],
+    };
+
+    let provider_reserved = match provider_reserved {
+        Some(reserved) if reserved.len() != PROVIDER_RESERVED_LENGTH => {
+            return Err(format!(
+                "When provided, there must be {} reserved bytes for resource provider metadata.",
+                PROVIDER_RESERVED_LENGTH
+            ));
+        }
+        Some(reserved) => reserved,
+        None => &[0; PROVIDER_RESERVED_LENGTH],
+    };
+
+    let base64_encoded_signature = if customer_managed_key {
+        base64_encoded_signature.to_uppercase()
+    } else {
+        base64_encoded_signature.to_lowercase()
+    };
+
+    let mut key = String::new();
+
+    loop {
+        let key_length_in_bytes = 66;
+        let mut key_bytes = vec![0; key_length_in_bytes];
+
+        if let Some(test_char) = test_char {
+            key = format!("{:86}==", test_char.to_string().repeat(86));
+        } else {
+            let mut rng = rand::thread_rng();
+            rng.fill_bytes(&mut key_bytes);
+        }
+
+        let key_string = general_purpose::STANDARD.encode(&key_bytes);
+        key_bytes = general_purpose::STANDARD.decode(&key_string).unwrap();
```

Comment:
@michaelcfanning This is a point where the Rust implementation differs from C#. I see that the output of the C# `GenerateCommonAnnotatedTestKey` is 85 characters, and not 88. However, the spec requires HIS v2 keys to be 66 bytes (or 88 characters). That's what I follow here. No trimming the string down to 86 characters and tagging two == in the end.